### PR TITLE
core.dtypes.generic --> cython

### PR DESCRIPTION
**Pull request by jbrockmendel**
Implement `ABCFoo` classes in cython.

Small optimization of the `typ in comps` check.  When `comps` is a singleton, this just becomes `typ == comp`.  I considered and decided against trying to take advantage of string interning to make it `typ is comp`.

Naturally asv is refusing to run (current hypothesis is it has something to do with a new directory existing), but in %timeit tests its an across-the-board speedup ~50%.

A new asv file is included in case someone else wants to try running it.

### CODE REVIEW

File: pandas/_libs/dtypes/generic.pyx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# cython: profile=False
+# cython: linetrace=False
+# distutils: define_macros=CYTHON_TRACE=0
+# distutils: define_macros=CYTHON_TRACE_NOGIL=0
+"""define generic base classes for pandas objects"""
+
+
+# define abstract base classes to enable isinstance type checking on our
+# objects
+def create_pandas_abc_type(name, attr, comp):
```
Comment (gfyoung) [inline pandas/_libs/dtypes/generic.pyx]: Where did this function live before?  Should delete that definition too.

Comment (gfyoung) [reply pandas/_libs/dtypes/generic.pyx]: Also, let's doc-string too while we're at it.

**Comment by jreback** [comment]
pls add an asv for each of these functions
then report the results

**Comment by jreback** [comment]
i seen u added the asv

pls run it / you need to figure how to run these stably

File: setup.py [hunk 2]
```diff
@@ -343,6 +343,7 @@ class CheckSDist(sdist_class):
                  'pandas/_libs/parsers.pyx',
                  'pandas/_libs/tslibs/timezones.pyx',
                  'pandas/_libs/tslibs/frequencies.pyx',
+                 'pandas/_libs/dtypes/generic.pyx',
```
Comment (jreback) [inline setup.py]: this will fail install 
you need to add in the list of installed dirs

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> you need to add in the list of installed dirs

Yep, that fixed the asv problem.

```
asv continuous -f 1.1 -E virtualenv master HEAD -b generic_dtypes
[...]
       before           after         ratio
     [21a38008]       [2addb85f]
-       201±0.8μs          143±1μs     0.71  generic_dtypes.GenericDtypes.time_isinstance

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

File: asv_bench/benchmarks/generic_dtypes.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import pandas as pd
+
+from pandas.core.dtypes import generic
+
+abcs = [getattr(generic, name) for name in dir(generic)
+        if name.startswith('ABC')]
+
+
+class GenericDtypes(object):
+    def setup(self):
```
Comment (jreback) [inline asv_bench/benchmarks/generic_dtypes.py]: you need a goal time

you need separate asv for each isinstance there is a way of doing this generically so it matrixes the tests
look in the other benches i think for bottleneck

**Review by jreback — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/generic_dtypes.py (hunk 3 above)
Comment (jbrockmendel) [reply asv_bench/benchmarks/generic_dtypes.py]: What constitutes a reasonable goal time?

Comment (jreback) [reply asv_bench/benchmarks/generic_dtypes.py]: i think if it's set too high it doesn't show the significantly changes properly 
have a look in asv docs as to actual purpose

File: asv_bench/benchmarks/generic_dtypes.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            i64index,
+            idx,
+            pd.Index(pd.compat.range(144, 169)),
+            mi,
+            pd.Categorical(['Do', 'Re', 'Mi', 'Fa']),
+            pd.CategoricalIndex(['Do'] * 5,
+                                categories=['Do', 'Re', 'Mi', 'Fa']),
+        ]
+
+    def time_isinstance_ABCCategorical(self):
```
Comment (jreback) [inline asv_bench/benchmarks/generic_dtypes.py]: what i mean is you don't actually write the individual tests - asv does this matrix for you 
it's way less code

File: pandas/_libs/dtypes/generic.pyx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    define abstract base classes to enable isinstance type checking on our
+    objects
+    """
+    comp = set(comp)
+
+    if len(comp) == 1:
+        comp = list(comp)[0]
+        @classmethod
```
Comment (jreback) [inline pandas/_libs/dtypes/generic.pyx]: does this actually matter? it's just more code 
i would leave the original

also when doing changes like this don't change code

very hard to detect what u changed

**Review by jreback — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/generic_dtypes.py (hunk 3 above)
Comment (jbrockmendel) [reply asv_bench/benchmarks/generic_dtypes.py]: Looks like this param adjusts the number of iterations to target the total runtime.  Default is 0.1

File: asv_bench/benchmarks/generic_dtypes.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    pd.Index(pd.compat.range(144, 169)),
+    mi,
+    pd.Categorical(['Do', 'Re', 'Mi', 'Fa']),
+    pd.CategoricalIndex(['Do'] * 5,
+                        categories=['Do', 'Re', 'Mi', 'Fa']),
+    ]
+
+
+class GenericDtypes(object):
+    goal_time = .0821
```
Comment (jreback) [inline asv_bench/benchmarks/generic_dtypes.py]: rather than do this in the module, pls use the setup method

**Review by jreback — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/generic_dtypes.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+import pandas as pd
+
+from pandas.core.dtypes import generic
+
+abcs = [getattr(generic, name) for name in dir(generic)
```
Comment (jreback) [inline asv_bench/benchmarks/generic_dtypes.py]: move down to the class

File: pandas/_libs/dtypes/generic.pyx (hunk 5 above)
Comment (jreback) [reply pandas/_libs/dtypes/generic.pyx]: pls respond here

**Comment by jreback** [comment]
can u show the result of the asv.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
Closing this to de-clutter the queue, may circle back to it when things calm down.

**Pull request closed without merging**